compiler/rustc_middle/src/mir/mod.rs

```diff
@@ -2605,9 +2605,30 @@ pub enum Rvalue<'tcx> {
 static_assert_size!(Rvalue<'_>, 40);
 
 impl<'tcx> Rvalue<'tcx> {
+    /// Returns true if evaluating rvalue can have side effects.
     #[inline]
-    pub fn is_pointer_int_cast(&self) -> bool {
-        matches!(self, Rvalue::Cast(CastKind::PointerAddress, _, _))
+    pub fn may_have_side_effects(&self) -> bool {
+        match self {
+            // Pointer to int casts may be side-effects due to exposing the provenance.
+            // While the model is undecided, we should be conservative. See
+            // <https://www.ralfj.de/blog/2022/04/11/provenance-exposed.html>
+            Rvalue::Cast(CastKind::PointerAddress, _, _) => true,
+
+            Rvalue::Use(_)
```

Comment:
`Use` can potentially mutate memory if it uses an `Operand::Move` (which is proposed to de-initialize the memory it reads from).
Removing a `Move` with unused result is still okay, but e.g. reordering it with other reads/writes is not if they might overlap.